// os/os_state.js
/*
   This file is part of UNIX Guide for the Perplexed project.
   Copyright (C) 2014 by Assaf Gordon <assafgordon@gmail.com>
   Released under GPLv3 or later, with the following addition:

     As additional permission under GNU GPL version 3 section 7, you
     may distribute non-source (e.g., minimized or compacted) forms of
     that code without the copy of the GNU GPL normally required by
     section 4, provided you include this license notice and a URL
     through which recipients can access the Corresponding Source.

   See: https://www.gnu.org/philosophy/javascript-trap.html
*/

/*
Kernel/Operating-System Encapsulation.

This module contains functions which could be consdered as
kernel system-calls which return global system information.
*/
"use strict";

var time_utils = require("utils/time_utils");

module.exports = OperatingSystem;


function OperatingSystemException(message) {
		this.message = message ;
		this.name = "OperatingSystemException";
}
OperatingSystemException.prototype = new Error();
OperatingSystemException.prototype.constructor = OperatingSystemException;

function OperatingSystem() {

	this.last_pid = 0;
	function allocate_pid()
	{
		this.last_pid++;
		return this.last_pid;
	}

	/* Similar to time(2) */
	function time()
	{
		return time_utils.get_sys_time();
	}

	/* Similar to uname(2) */
	function uname()
	{
		return {
			"sysname" : "agnostic",
			"nodename" : "FOOBAR",
			"release" : "0.0.1", // "kernel version"
			"version" : "0.0.1", //TODO: get version from GIT
			"machine" : "JavaScript" // TODO: Use Browser or UserAgent Detection?
			};
	}

	this.time = time;
	this.uname = uname;
	this.allocate_pid = allocate_pid ;
	return this;
}
