// utils/object_utils.js
/*
   This file is part of UNIX Guide for the Perplexed project.
   Copyright (C) 2014 by Assaf Gordon <assafgordon@gmail.com>
   Released under GPLv3 or later, with the following addition:

     As additional permission under GNU GPL version 3 section 7, you
     may distribute non-source (e.g., minimized or compacted) forms of
     that code without the copy of the GNU GPL normally required by
     section 4, provided you include this license notice and a URL
     through which recipients can access the Corresponding Source.

   See: https://www.gnu.org/philosophy/javascript-trap.html
*/

/*
Generate Helper Functions.

Usage Example:
  var ob_utils_wrapper = require('utils/object_utils');
  var ou = ob_utils_wrapper.ob_utils;
  if (ou.IsNumber('foo')) { ... }


TODO:
1. Replace this module with UnderscoreJS ?
2. IsObject and IsArray are limited (perhaps even incorrect),
   according to http://tobyho.com/2011/01/28/checking-types-in-javascript/

 */

module.exports = {
	ObjectException: ObjectException,
	ob_utils: (new ObjectUtils())    //NOTE: this is a singleton, not a class
};



function ObjectException(message,data) {
		this.message = message ;
		this.name = "ObjectException";
		this.data = data ;
		this.toString = function() {
			if (data)
				return this.name + ": " + message + " (data = " +
					JSON.stringify(data) + ")";
			return this.name + ": " + message;
		}
	}
ObjectException.prototype = new Error();
ObjectException.prototype.constructor = ObjectException;


function ObjectUtils() {

this.IsObjectEmpty = function(obj) {
	return (Object.keys(obj).length === 0);
}

this.GetObjectType = function(obj) {
	var s = Object.prototype.toString.call(obj);
	return s;
}
this.IsNumber = function(obj) {
	var s = this.GetObjectType(obj);
	return (  s === "[object Number]");
}
this.VerifyNumber = function (obj) {
	if (!this.IsNumber(obj))
		throw new ObjectException("variable is not a number");
}
this.IsInteger = function(obj) {
	return (this.IsNumber(obj)
		&&
		(Math.floor(obj) == obj));
}
this.VerifyInteger = function (obj) {
	if (!this.IsInteger(obj))
		throw new ObjectException("variable is not an Integer");
}


this.IsBoolean = function(obj) {
	var s = this.GetObjectType(obj);
	return (  s === "[object Boolean]");
}
this.VerifyBoolean = function (obj) {
	if (!this.IsBoolean(obj))
		throw new ObjectException("variable is not boolean");
}
this.IsString = function(obj) {
	var s = this.GetObjectType(obj);
	return (  s === "[object String]");
}
this.VerifyString = function (obj) {
	if (!this.IsString(obj))
		throw new ObjectException("variable is not string");
}

/* returns TRUE if a variable holds a value that
   can be converted to a floating point value, and DOES NOT
   contain any extra characters. The only allowed exception
   is leading white-space.
   This is in contrast to javascript's parseFloat() which allows
   trailing characters.
   Example:
	parseFloat(" 5.5 apples") === 5.5
   but
        IsStrictFloatNumber(" 5.5 apples") === false

   This is similar to 'strtold(3)' function, WITH verifying the value
   of 'endptr' after conversion.
*/
this.IsStrictFloatValue = function(num) {
	if (this.IsNumber(num))
		return true;
	if (this.IsString(num)&&
	     (/^[ \t]*[-+]?[0-9]*\.?[0-9]+$/.test(num) ||
	      /^[ \t]*[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$/.test(num)))
		return true;
	return false;
}

/* returns TRUE if a variable holds a value that
   can be converted to an integer (as decimal value), and DOES NOT
   contain any extra characters. The only allowed exception
   is leading white-space.
   This is in contrast to javascript's parseInt() which allows
   trailing characters.
   Example:
	parseInt(" 5.5 apples") === 5
   but
        IsStrictDecimalIntegerValue(" 5.5 apples") === false

   This is similar to 'strtoul(3)' function, WITH verifying the value
   of 'endptr' after conversion.
*/
this.IsStrictDecimalIntegerValue = function(num) {
	if (this.IsInteger(num))
		return true;
	if (this.IsString(num) && /^[ \t]*[-+]?[0-9]+$/.test(num))
		return true;
	return false;
}


/* A Helper function, returns true of the given parameter is an object */
this.IsObject = function(obj) {
	var s = this.GetObjectType(obj);
	return (  s === "[object Object]");
}
this.VerifyObject = function (obj) {
	if (!this.IsObject(obj))
		throw new ObjectException("'obj' is not a valid object", obj);
}


/* A Helper function, returns true of the given parameter is an array */
this.IsArray = function (obj) {
	var s = this.GetObjectType(obj);
	return (  s === "[object Array]");
}

this.VerifyArray = function (obj) {
	if (!this.IsArray(obj))
		throw new ObjectException("'obj' is not a valid array", obj);
}

/* A Helper function to verify that a given object has only the allowed keys.
   The function throws 'ObjectException' if it finds a disallowed key.

   Example:
	a = { "foo" : "bar", "baz" : [1,2,3,4,5] };
	VerifyAllowedKeys(a, ["foo","baz"]); //Will succeed
	VerifyAllowedKeys(a, ["foo"]); //Will throw exception: "baz" is not an allowed key
*/
this.VerifyAllowedKeys = function(obj, allowed_keys_array) {
	this.VerifyObject(obj);

	/* Create a hash of allowed keys */
	var allowed_keys = {};
	for (var i in allowed_keys_array) {
		allowed_keys[allowed_keys_array[i]] = 1 ;
	}

	/* Iterate over keys in obj */
	for (var key in obj) {
		var ok = (key in allowed_keys);
		if (! (key in allowed_keys) )
			throw new ObjectException(
				":'obj' contains disallowed key '" + key + "'" +
				" allowed-keys = '" + allowed_keys_array.toString() + "'" ,obj);
	}
}

/* A Helper function to verify a hash contains just one key.
   The function throws if there are ZERO keys, or two-or-more keys.

   Example:
	VerifyOneKey( { "foo": [1,2,3,4] } ); // OK
	VerifyOneKey( { } ) ; // throws
	VerifyOneKey( { "foo" : 4, "bar": 43 } ) ; //throws
*/
this.VerifyOneKey = function(obj) {
	this.VerifyObject(obj);

	if (Object.keys(obj).length==0)
		throw new ObjectException("'obj' is empty", obj);
	if (Object.keys(obj).length>1)
		throw new ObjectException("'obj' contains more than one key", obj);
}

/* A Helper function to return the name of the one key in an object.
   The function will throw if the objecct doesn't have keys, or have more
   than more key.

   Example:
	GetOneKey( { "foo" : [1,2,3,4] } ); //returns "foo"
	GetOneKey( { } ); //throws
	GetOneKey( { "foo": 3, "bar": [1,2,3,4] } ); //throws.
*/
this.GetOneKey = function(obj) {
	this.VerifyOneKey(obj);
	for (var key in obj) {
		return key;
	}
}

}
