// utils/object_utils.js
/*
   This file is part of UNIX Guide for the Perplexed project.
   Copyright (C) 2014 by Assaf Gordon <assafgordon@gmail.com>
   Released under GPLv3 or later, with the following addition:

     As additional permission under GNU GPL version 3 section 7, you
     may distribute non-source (e.g., minimized or compacted) forms of
     that code without the copy of the GNU GPL normally required by
     section 4, provided you include this license notice and a URL
     through which recipients can access the Corresponding Source.

   See: https://www.gnu.org/philosophy/javascript-trap.html
*/

/*
Generate Helper Functions.

Usage Example:
  var ob_utils_wrapper = require('utils/object_utils');
  var ou = ob_utils_wrapper.ob_utils;
  if (ou.IsNumber('foo')) { ... }


TODO:
1. Replace this module with UnderscoreJS ?
2. IsObject and IsArray are limited (perhaps even incorrect),
   according to http://tobyho.com/2011/01/28/checking-types-in-javascript/

 */

module.exports = new ObjectUtils();    //NOTE: this is a singleton, not a class

function ObjectUtils() {

function IsObjectEmpty(obj) {
	return (Object.keys(obj).length === 0);
}
this.IsObjectEmpty = IsObjectEmpty;

function GetObjectType(obj) {
	var s = Object.prototype.toString.call(obj);
	return s;
}
this.GetObjectType = GetObjectType;

function IsNumber(obj) {
	var s = GetObjectType(obj);
	return (  s === "[object Number]");
}
this.IsNumber = IsNumber;

function VerifyNumber(obj) {
	if (!IsNumber(obj))
		throw new TypeError("variable is not a number");
}
this.VerifyNumber = VerifyNumber;

function IsInteger(obj) {
	return (IsNumber(obj)
		&&
		(Math.floor(obj) == obj));
}
this.IsInteger = IsInteger;

function VerifyInteger(obj) {
	if (!IsInteger(obj))
		throw new TypeError("variable is not an Integer");
}
this.VerifyInteger = VerifyInteger;


function IsBoolean(obj) {
	var s = GetObjectType(obj);
	return (  s === "[object Boolean]");
}
this.IsBoolean = IsBoolean;


function VerifyBoolean(obj) {
	if (!IsBoolean(obj))
		throw new TypeError("variable is not boolean");
}
this.VerifyBoolean = VerifyBoolean;

function IsString(obj) {
	var s = GetObjectType(obj);
	return (  s === "[object String]");
}
this.IsString = IsString;

function VerifyString(obj) {
	if (!IsString(obj))
		throw new TypeError("variable is not string");
}
this.VerifyString = VerifyString;

/* returns TRUE if a variable holds a value that
   can be converted to a floating point value, and DOES NOT
   contain any extra characters. The only allowed exception
   is leading white-space.
   This is in contrast to javascript's parseFloat() which allows
   trailing characters.
   Example:
	parseFloat(" 5.5 apples") === 5.5
   but
        IsStrictFloatNumber(" 5.5 apples") === false

   This is similar to 'strtold(3)' function, WITH verifying the value
   of 'endptr' after conversion.
*/
function IsStringFloatValue(num) {
	if (IsNumber(num))
		return true;
	if (IsString(num)&&
	     (/^[ \t]*[-+]?[0-9]*\.?[0-9]+$/.test(num) ||
	      /^[ \t]*[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$/.test(num)))
		return true;
	return false;
}
this.IsStrictFloatValue = IsStringFloatValue;

/* returns TRUE if a variable holds a value that
   can be converted to an integer (as decimal value), and DOES NOT
   contain any extra characters. The only allowed exception
   is leading white-space.
   This is in contrast to javascript's parseInt() which allows
   trailing characters.
   Example:
	parseInt(" 5.5 apples") === 5
   but
        IsStrictDecimalIntegerValue(" 5.5 apples") === false

   This is similar to 'strtoul(3)' function, WITH verifying the value
   of 'endptr' after conversion.
*/
function IsStrictDecimalIntegerValue(num) {
	if (IsInteger(num))
		return true;
	if (IsString(num) && /^[ \t]*[-+]?[0-9]+$/.test(num))
		return true;
	return false;
}
this.IsStrictDecimalIntegerValue = IsStrictDecimalIntegerValue;


/* A Helper function, returns true of the given parameter is an object */
function IsObject(obj) {
	var s = GetObjectType(obj);
	return (  s === "[object Object]");
}
this.IsObject = IsObject;


function VerifyObject(obj) {
	if (!IsObject(obj))
		throw new TypeError("'obj' is not a valid object: " + JSON.stringify(obj));
}
this.VerifyObject = VerifyObject;


/* A Helper function, returns true of the given parameter is an array */
function IsArray(obj) {
	var s = GetObjectType(obj);
	return (  s === "[object Array]");
}
this.IsArray = IsArray;

function VerifyArray(obj) {
	if (!IsArray(obj))
		throw new TypeError("'obj' is not a valid array:" + JSON.stringify(obj));
}
this.VerifyArray = VerifyArray;

/* A Helper function to verify that a given object has only the allowed keys.
   The function throws 'TypeError' if it finds a disallowed key.

   Example:
	a = { "foo" : "bar", "baz" : [1,2,3,4,5] };
	VerifyAllowedKeys(a, ["foo","baz"]); //Will succeed
	VerifyAllowedKeys(a, ["foo"]); //Will throw exception: "baz" is not an allowed key
*/
function VerifyAllowedKeys(obj, allowed_keys_array) {
	VerifyObject(obj);

	/* Create a hash of allowed keys */
	var allowed_keys = {};
	for (var i in allowed_keys_array) {
		allowed_keys[allowed_keys_array[i]] = 1 ;
	}

	/* Iterate over keys in obj */
	for (var key in obj) {
		var ok = (key in allowed_keys);
		if (! (key in allowed_keys) )
			throw new TypeError(
				":'obj' contains disallowed key '" + key + "'" +
				" allowed-keys = '" + allowed_keys_array.toString() + "': " + JSON.stringify(obj));
	}
}
this.VerifyAllowedKeys = VerifyAllowedKeys;

/* A Helper function to verify a hash contains just one key.
   The function throws if there are ZERO keys, or two-or-more keys.

   Example:
	VerifyOneKey( { "foo": [1,2,3,4] } ); // OK
	VerifyOneKey( { } ) ; // throws
	VerifyOneKey( { "foo" : 4, "bar": 43 } ) ; //throws
*/
function VerifyOneKey(obj) {
	VerifyObject(obj);

	if (Object.keys(obj).length==0)
		throw new TypeError("'obj' is empty");
	if (Object.keys(obj).length>1)
		throw new TypeError("'obj' contains more than one key: " +  JSON.stringify(obj));
}
this.VerifyOneKey = VerifyOneKey;

/* A Helper function to return the name of the one key in an object.
   The function will throw if the objecct doesn't have keys, or have more
   than more key.

   Example:
	GetOneKey( { "foo" : [1,2,3,4] } ); //returns "foo"
	GetOneKey( { } ); //throws
	GetOneKey( { "foo": 3, "bar": [1,2,3,4] } ); //throws.
*/
function GetOneKey(obj) {
	VerifyOneKey(obj);
	for (var key in obj) {
		return key;
	}
}
this.GetOneKey = GetOneKey;

}
